### PR TITLE
HINT: Inlay parameter type hints in struct literals implemented

### PR DESCRIPTION
**Pull request by dem1tris**

### CODE REVIEW

**Comment by vlad20012** [comment]
Do we really need this? I think these hints are mostly redundant. We don't, for example, specially show function argument types in hints. In case of struct, argument names are already provided in syntax, and I think it's enough. 
image

**Comment by dem1tris** [comment]
@vlad20012 I thought about this hint while creating literal of one dummy struct with couple of dummy fields with types I actually don't remember. And it seemed like "type something here, I won't tell what exactly, but I will highlight type mismatch, If you make a mistake". I suggested that we need an easier way than "go to declaration". By the way, it can be disabled, if redundant

**Comment by vlad20012** [comment]
Compare it with a case of function call. We don't show types too. But you can ctrl+P on it and see popup with function signature.
image

**Comment by dem1tris** [comment]
@vlad20012 Yes, but on struct literal we can't use Ctrl+P) Maybe it's worth turning it off by default? Or make it disappear, when correct value typed?

**Comment by vlad20012** [comment]
Maybe it's better to make ctrl+P work with struct literals? =)

**Comment by sirgl** [comment]
We can make an option to make user decide if he wants to see it or not with false default value.

File: src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt [hunk 1]
```diff
@@ -112,6 +116,29 @@ enum class HintType(desc: String, enabled: Boolean) {
...
+    /* In case of fields without any expressions
+       PSI is destructed in this place, so only hints before error element will be shown */
+    STRUCT_LITERAL_TYPE_HINT("Show type hints for fields in struct literal", true) {
+        override fun provideHints(elem: PsiElement): List<InlayInfo> {
+            val element = elem as? RsStructLiteral ?: return emptyList()
+            val declaration = element.path.reference.resolve() as? RsFieldsOwner ?: return emptyList()
+            val subst = element.type.typeParameterValues
+            val hints = element.structLiteralBody.structLiteralFieldList.mapNotNull { field ->
+                (declaration.namedFields
+                    .find { it.name.equals(field.referenceName) }
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt]: You can use `it.reference.resolve() as? RsFieldDecl` instead

File: src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt [hunk 2]
```diff
@@ -112,6 +116,29 @@ enum class HintType(desc: String, enabled: Boolean) {
...
 
         override fun isApplicable(elem: PsiElement): Boolean = elem is RsForExpr
+    },
+
+    /* In case of fields without any expressions
+       PSI is destructed in this place, so only hints before error element will be shown */
+    STRUCT_LITERAL_TYPE_HINT("Show type hints for fields in struct literal", true) {
+        override fun provideHints(elem: PsiElement): List<InlayInfo> {
+            val element = elem as? RsStructLiteral ?: return emptyList()
+            val declaration = element.path.reference.resolve() as? RsFieldsOwner ?: return emptyList()
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt]: ... and so `declaration` var will become unused

File: src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt [hunk 3]
```diff
@@ -112,6 +116,29 @@ enum class HintType(desc: String, enabled: Boolean) {
...
+            val subst = element.type.typeParameterValues
+            val hints = element.structLiteralBody.structLiteralFieldList.mapNotNull { field ->
+                (declaration.namedFields
+                    .find { it.name.equals(field.referenceName) }
+                    ?.typeReference?.type?.substitute(subst)
+                    ?.shortPresentableText
+                    ?: return@mapNotNull null) to (
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt]: it is very unobvious construction. Let's extract left and right parts of `to` expression to variables.

File: src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt [hunk 4]
```diff
@@ -112,6 +114,29 @@ enum class HintType(desc: String, enabled: Boolean) {
...
+            val hints = element.structLiteralBody.structLiteralFieldList.mapNotNull { field ->
+                val text = (field.reference.resolve() as? RsFieldDecl)
+                    ?.typeReference?.type?.substitute(subst)
+                    ?.shortPresentableText
+                    ?: return@mapNotNull null
+                // in case of incomplete field, COLON will be the sibling, not the child
+                val offset =
+                    (field.colon ?: if (field.nextSibling.elementType == RsElementTypes.COLON) field.nextSibling
+                    else field.identifier).textRange.endOffset
+                text to offset
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/hints/RsInlayParameterHintsProvider.kt]: There was the second comment. You can (I guess) write here directly InlayInfo(text, offset), and then just return `hints` without additional mapping

**Comment by vlad20012** [comment]
I finally decided that we do not need it. We show parameter name hints at function calls, but we don't show their types. But if you want you can see parameter types via ctrl+P. In the case of struct fields, we have their names in syntax already, and I don't see much difference between function parameters and struct fields here. If names are already "shown", I think we don't have to show type hints in this special case. To be uniform we should provide ctrl+P on struct literals.

**Pull request closed without merging**